adding ecr policy resource

enderv opened the pull request:
Signed-off-by: chris <cvodak@ea.com>

<!--
Thank you for helping to improve Crossplane!

Please read through https://git.io/fj2m9 if this is your first time opening a
Crossplane pull request. Find us in https://slack.crossplane.io/messages/dev if
you need any help contributing.
-->

### Description of your changes

<!--
Briefly describe what this pull request does. Be sure to direct your reviewers'
attention to anything that needs special consideration.

We love pull requests that resolve an open Crossplane issue. If yours does, you
can uncomment the below line to indicate which issue your PR fixes, for example
"Fixes #500":

-->
Fixes #535
Adds support for ECR Policies by copying the same approach as S3 Bucket policy where you can either do raw string or a struct based policy. 

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested
Run automated tests. Tested by adding example policy to a repo created using a repo resource. Modified the policy to make sure changes went through.


<!--
Before reviewers can be confident in the correctness of this pull request, it
needs to tested and shown to be correct. Briefly describe the testing that has
already been done or which is planned for this change.
-->

[contribution process]: https://git.io/fj2m9

chlunde commented on examples/ecr/respositorypolicy-json.yaml:
```diff
@@ -0,0 +1,23 @@
+  forProvider:
+    region: us-east-1
+    repositoryName: example
+    RawPolicy:  |
```
```suggestion
    rawPolicy:  |
```

Just skimmed the PR and noticed this, I have not reviewed it :)

enderv replied:
doh, nice catch.

hasheddan commented on pkg/controller/ecr/repositorypolicy/controller.go:
```diff
@@ -0,0 +1,209 @@
+	policyData, err := e.formatRepositoryPolicy(cr)
+
+	if err != nil {
+		return managed.ExternalObservation{}, awsclient.Wrap(resource.Ignore(ecr.IsPolicyNotFoundErr, err), errGet)
```
I don't think we would ever hit `ecr.IsPolicyNotFoundErr` here right?

enderv replied:
yep. will fix

enderv commented on pkg/controller/ecr/repositorypolicy/controller.go:
```diff
@@ -0,0 +1,209 @@
+	}
+	cr.Status.AtProvider.PolicyText = *policyData
+
+	if !cmp.Equal(*policyData, *response.PolicyText) {
```
thats my understanding.

hasheddan replied:
@enderv I think are some places where we do this with the assumption that it will never be `nil`. I personally prefer to check that it is not `nil` as, though it wouldn't be great by SDK maintainers to change this behavior, it could be changed and we would then have a potential panic here. What do you think?

enderv replied:
sounds good to me. On it

enderv replied:
Oh actually already got this covered with my lateinitialization change have a nil check in the function there and now just comparing the `RepositoryPolicyParameters`

hasheddan replied:
Aren't we still deferencing before late init?

https://github.com/crossplane/provider-aws/pull/573/files#diff-b54d7ac2e4ab0e950fd8fb8c791b11196da72121140174398d12c779bc643624R116

enderv replied:
:eyes: so we are. on it.

enderv commented on pkg/controller/ecr/repositorypolicy/controller.go:
```diff
@@ -0,0 +1,209 @@
+	if err != nil {
+		return managed.ExternalObservation{}, awsclient.Wrap(resource.Ignore(ecr.IsPolicyNotFoundErr, err), errGet)
+	}
+	cr.Status.AtProvider.PolicyText = *policyData
```
Oh good point. yep will change.

enderv commented on pkg/controller/ecr/repositorypolicy/controller.go:
```diff
@@ -0,0 +1,209 @@
+			return managed.ExternalObservation{}, awsclient.Wrap(err, errSpecUpdate)
+		}
+	}
+
+	cr.SetConditions(xpv1.Available())
```
Doh, had to double check, thought there wasn't any but looks like two fields will get em in.

enderv commented:
@hasheddan think I got everything, not sure how you normally do this just resolved as I went through.

HerrmannHinz commented:
niceone @enderv

muvaf commented on apis/ecr/v1alpha1/policy_types.go:
```diff
@@ -0,0 +1,272 @@
+/*
+Copyright 2020 The Crossplane Authors.
```
```suggestion
Copyright 2021 The Crossplane Authors.
```

in all files.

muvaf commented on apis/ecr/v1alpha1/referencers.go:
```diff
@@ -0,0 +1,106 @@
+		Extract:      reference.ExternalName(),
+	})
+	if err != nil {
+		return errors.Wrap(err, "spec.forProvider.RepositoryName")
```
```suggestion
		return errors.Wrap(err, "spec.forProvider.repositoryName")
```

muvaf commented on apis/ecr/v1alpha1/referencers.go:
```diff
@@ -0,0 +1,106 @@
+	mg.Spec.ForProvider.RepositoryNameRef = rsp.ResolvedReference
+
+	// Resolve spec.forProvider.userName
+	if mg.Spec.ForProvider.Policy != nil && mg.Spec.ForProvider.Policy.Statements != nil {
```
```suggestion
	if mg.Spec.ForProvider.Policy != nil {
```

`for` would be no-op if array is `nil` or has no elements.

muvaf commented on apis/ecr/v1alpha1/referencers.go:
```diff
@@ -0,0 +1,106 @@
+		return nil
+	}
+	for i := range principal.AWSPrincipals {
+		if principal.AWSPrincipals[i].IAMUserARNRef != nil || principal.AWSPrincipals[i].IAMUserARNSelector != nil {
```
```suggestion
```
This `nil` check is taken care of by the resolver.

muvaf commented on apis/ecr/v1alpha1/referencers.go:
```diff
@@ -0,0 +1,106 @@
+				Extract:      identityv1alpha1.IAMUserARN(),
+			})
+			if err != nil {
+				return errors.Wrap(err, fmt.Sprintf("spec.forProvider.statement[%d].principal.aws[%d].IAMUserARN", statementIndex, i))
```
```suggestion
				return errors.Wrap(err, fmt.Sprintf("spec.forProvider.statements[%d].principal.awsPrincipals[%d].iamUserArn", statementIndex, i))
```

muvaf commented on apis/ecr/v1alpha1/referencers.go:
```diff
@@ -0,0 +1,106 @@
+				Extract:      identityv1beta1.IAMRoleARN(),
+			})
+			if err != nil {
+				return errors.Wrap(err, fmt.Sprintf("spec.forProvider.statement[%d].principal.aws[%d].IAMRoleArn", statementIndex, i))
```
```suggestion
				return errors.Wrap(err, fmt.Sprintf("spec.forProvider.statements[%d].principal.awsPrincipals[%d].iamRoleArn", statementIndex, i))
```

muvaf commented on examples/ecr/repositorypolicy.yaml:
```diff
@@ -0,0 +1,23 @@
+spec:
+  forProvider:
+    region: us-east-1
+    repositoryName: example
```
```suggestion
    repositoryNameRef:
      name: example
```

muvaf commented on pkg/controller/ecr/repositorypolicy/controller.go:
```diff
@@ -0,0 +1,214 @@
+	}).Send(ctx)
+
+	if err != nil {
+		if ecr.IsRepoNotFoundErr(err) {
+			return managed.ExternalObservation{}, nil
+		}
```
```suggestion
```

I think that's a valid error case - users cannot set policy of a repository that doesn't exist. We ignore `PolicyNotFoundErr` because this controller will create the policy.

muvaf commented on pkg/controller/ecr/repositorypolicy/controller.go:
```diff
@@ -0,0 +1,214 @@
+
+	current := cr.Spec.ForProvider.DeepCopy()
+	ecr.LateInitializeRepositoryPolicy(&cr.Spec.ForProvider, response)
+	if !cmp.Equal(current, &cr.Spec.ForProvider) {
+		if err := e.kube.Update(ctx, cr); err != nil {
+			return managed.ExternalObservation{}, awsclient.Wrap(err, errSpecUpdate)
+		}
+	}
```
```suggestion
```

You can return `ResourceLateInitialized: !cmp.Equal(current, &cr.Spec.ForProvider)` now with the latest crossplane-runtime.

muvaf commented on pkg/controller/ecr/repositorypolicy/controller.go:
```diff
@@ -0,0 +1,214 @@
+		return managed.ExternalCreation{}, errors.New(errUnexpectedObject)
+	}
+
+	cr.Status.SetConditions(xpv1.Creating())
```
```suggestion
```

The generic reconciler will set this now automatically.

muvaf commented on pkg/controller/ecr/repositorypolicy/controller.go:
```diff
@@ -0,0 +1,214 @@
+	}
+
+	cr.Status.SetConditions(xpv1.Creating())
+	if err := e.kube.Status().Update(ctx, cr); err != nil {
+		return managed.ExternalCreation{}, errors.Wrap(err, errStatusUpdate)
+	}
```
```suggestion
```

muvaf commented on pkg/controller/ecr/repositorypolicy/controller.go:
```diff
@@ -0,0 +1,214 @@
+		return errors.New(errUnexpectedObject)
+	}
+
+	cr.Status.SetConditions(xpv1.Deleting())
```
```suggestion
```

The generic reconciler will set this now automatically.

muvaf commented on pkg/controller/ecr/repositorypolicy/controller.go:
```diff
@@ -0,0 +1,214 @@
+}
+
+// formatRepositoryPolicy parses and formats the RepositoryPolicy struct
+func (e *external) formatRepositoryPolicy(original *v1alpha1.RepositoryPolicy) (*string, error) {
```
```suggestion
func (e *external) rawPolicyData(original *v1alpha1.RepositoryPolicy) (string, error) {
```

nitpick: returning `string` could make it easier to work with. We'd use `awsclients.String(ref)` as usual whenever we send it to AWS. And `formatRepositoryPolicy` feels like there is an in-place change but what we do is mostly conversion from one format to another.

muvaf replied:
Also what do you think about moving this to `pkg/clients/ecr/repository_policy.go`?

enderv replied:
kk and i'll go ahead and move it as well

muvaf commented on pkg/controller/ecr/repositorypolicy/controller.go:
```diff
@@ -0,0 +1,214 @@
+		RepositoryName: cr.Spec.ForProvider.RepositoryName,
+		RegistryId:     cr.Spec.ForProvider.RegistryID,
+	}).Send(ctx)
+	if ecr.IsRepoNotFoundErr(err) {
+		return nil
+	}
```
```suggestion
```

muvaf commented on pkg/controller/ecr/repositorypolicy/controller.go:
```diff
@@ -0,0 +1,214 @@
+		return managed.ExternalUpdate{}, errors.Wrap(err, errUpdate)
+	}
+	_, err = e.client.SetRepositoryPolicyRequest(ecr.GenerateSetRepositoryPolicyInput(&cr.Spec.ForProvider, policyData)).Send(ctx)
+	if err != nil {
+		return managed.ExternalUpdate{}, awsclient.Wrap(err, errUpdate)
+	}
+	return managed.ExternalUpdate{}, nil
```
```suggestion
	return managed.ExternalUpdate{}, awsclient.Wrap(err, errUpdate)
```

muvaf commented on pkg/controller/ecr/repositorypolicy/controller.go:
```diff
@@ -0,0 +1,214 @@
+		return managed.ExternalCreation{}, errors.Wrap(err, errCreate)
+	}
+	_, err = e.client.SetRepositoryPolicyRequest(ecr.GenerateSetRepositoryPolicyInput(&cr.Spec.ForProvider, policyData)).Send(ctx)
+	if err != nil {
+		return managed.ExternalCreation{}, awsclient.Wrap(err, errCreate)
+	}
+	return managed.ExternalCreation{}, errors.Wrap(e.kube.Update(ctx, cr), errSpecUpdate)
```
```suggestion
	return managed.ExternalCreation{}, awsclient.Wrap(err, errCreate)
```

Is there a specific reason that requires us to issue a spec update?

enderv replied:
Thought there was but looking again can't seem to figure out why will remove that.

muvaf commented on apis/ecr/v1alpha1/doc.go:
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2020 The Crossplane Authors.
+Copyright 2021 The Crossplane Authors.
```
We use the current year in only the files created with that PR.

enderv replied:
oops

muvaf commented on apis/ecr/v1alpha1/referencers.go:
```diff
@@ -0,0 +1,105 @@
+		principal.AWSPrincipals[i].IAMUserARN = reference.ToPtrValue(rsp.ResolvedValue)
+		principal.AWSPrincipals[i].IAMUserARNRef = rsp.ResolvedReference
+
+		if principal.AWSPrincipals[i].IAMRoleARNRef != nil || principal.AWSPrincipals[i].IAMRoleARNSelector != nil {
```
```suggestion
```

muvaf commented on apis/ecr/v1alpha1/register.go:
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2020 The Crossplane Authors.
+Copyright 2021 The Crossplane Authors.
```
```suggestion
Copyright 2020 The Crossplane Authors.
```

muvaf commented on pkg/clients/ecr/repository_policy.go:
```diff
@@ -0,0 +1,236 @@
+	if r == nil {
+		return
+	}
+	in.RegistryID = r.RegistryId
```
```suggestion
	in.RegistryID = awsclient.LateInitializeStringPtr(in.RegistryID, r.RegistryId)
```

We can't override spec if it already exists.

muvaf commented on apis/ecr/v1alpha1/policy_types.go:
```diff
@@ -0,0 +1,272 @@
+// RepositoryPolicyParameters define the desired state of an AWS Elastic Container Repository
+type RepositoryPolicyParameters struct {
+
+	// Region is the region you'd like your Repository to be created in.
```
```suggestion
	// Region is the region you'd like your RepositoryPolicy to be created in.
```

muvaf reviewed (approved):
Looking pretty good, should be ready to merge after a few quick fixes and commit squashing. `IsUpToDate` logic seems a bit risky, have you tested to see whether it returns false negatives?

enderv commented:
@muvaf got the rest of those changes in and squashed. For the `IsUpToDate` did a few checks just changing ordering of the json to make sure equivalent came through correct. Added another test but if anything else in mind to check just let me know.

muvaf commented:
Thank you @enderv !!
